File: go-matt/protocol/messages.go
package protocol

import (
	"encoding/hex"
	"encoding/json"
	"log"
)

func Key_Exchange(pub []byte) []byte {
	pub_hex := hex.EncodeToString(pub)
	key := Key_Exch_Msg{KEY_EXCHANGE, pub_hex}

	msg, err := json.Marshal(key)
	if err != nil {
		log.Panicln("Error encoding public key")
		return []byte{}
	}

	return msg
}

func Hello(name string, identity_pub []byte, fingerprint []byte) []byte {
	pub_hex := hex.EncodeToString([]byte(identity_pub))
	fingerprint_hex := hex.EncodeToString([]byte(fingerprint))
	key := Hello_Msg{HELLO, name, pub_hex, fingerprint_hex}

	msg, err := json.Marshal(key)
	if err != nil {
		log.Panicln("Error encoding public key")
		return []byte{}
	}

	return msg
}

func DataMessage(nonce []byte, payload []byte) []byte {
	nonceHex := hex.EncodeToString(nonce)
	payloadHex := hex.EncodeToString(payload)

	msg, err := json.Marshal(Data_Msg{DATA, nonceHex, payloadHex})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}
	return msg
}

func FileListRequest() []byte {
	msg, err := json.Marshal(Msg{FILE_LIST_REQUEST})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}

	return msg
}

func FileListResponse(records []*FileRecord) []byte {
	msg, err := json.Marshal(File_List_Res_Msg{FILE_LIST_RESPONSE, records})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}

	return msg
}

func FileRequest(filename string) []byte {
	msg, err := json.Marshal(File_Req_Msg{FILE_REQUEST, filename})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}

	return msg
}

// def file_offer(filename: str, record: dict):
//     return {
//         "type": FILE_OFFER,
//         "filename": filename,
//         "record": record,
//     }

// def file_offer_response(filename: str, accepted: bool, message: str = ""):
//     return {
//         "type": FILE_OFFER_RESPONSE,
//         "filename": filename,
//         "accepted": accepted,
//         "message": message,
//     }

func FileChunk(filename string, data []byte, record *FileRecord, done bool) []byte {
	dataHex := hex.EncodeToString(data)

	msg, err := json.Marshal(File_Chunk_Msg{
		Type:     FILE_CHUNK,
		Filename: filename,
		Data:     dataHex,
		Record:   record,
		Done:     done,
	})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}

	return msg
}

func KeyMigration(newpub []byte, oldsig []byte, newsig []byte) []byte {
	newPubHex := hex.EncodeToString(newpub)
	oldSigHex := hex.EncodeToString(oldsig)
	newSigHex := hex.EncodeToString(newsig)

	msg, err := json.Marshal(Key_Migration_Msg{KEY_MIGRATION, newPubHex, oldSigHex, newSigHex})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}

	return msg
}

func ErrorMessage(message string) []byte {
	msg, err := json.Marshal(Error_Msg{ERROR, message})
	if err != nil {
		log.Panicln("Something went wrong", err)
		return []byte{}
	}

	return msg
}

File: python-client/protocol/messages.py
import json

from protocol.message_types import (
    DATA,
    ERROR,
    FILE_CHUNK,
    FILE_LIST_REQUEST,
    FILE_LIST_RESPONSE,
    FILE_OFFER,
    FILE_OFFER_RESPONSE,
    FILE_REQUEST,
    HELLO,
    KEY_CONFIRM,
    KEY_EXCHANGE,
    KEY_MIGRATION,
)


def key_exchange(pub_hex: str):
    return {
        "type": KEY_EXCHANGE,
        "pub": pub_hex,
    }


def hello(
    name: str,
    identity_pub_hex: str,
    fingerprint: str,
    signature_hex: str = "",
):
    message = {
        "type": HELLO,
        "name": name,
        "identity_pub": identity_pub_hex,
        "fingerprint": fingerprint,
    }
    if signature_hex:
        message["signature"] = signature_hex
    return message


def data_message(nonce_hex: str, payload_hex: str):
    return {
        "type": DATA,
        "nonce": nonce_hex,
        "payload": payload_hex,
    }


def key_confirm(echoed_identity_pub_hex: str, signer_identity_pub_hex: str, signature_hex: str):
    return {
        "type": KEY_CONFIRM,
        "echoed_identity_pub": echoed_identity_pub_hex,
        "signer_identity_pub": signer_identity_pub_hex,
        "signature": signature_hex,
    }


def file_list_request():
    return {"type": FILE_LIST_REQUEST}


def file_list_response(files: list):
    return {
        "type": FILE_LIST_RESPONSE,
        "files": files,
    }


def file_request(filename: str):
    return {
        "type": FILE_REQUEST,
        "filename": filename,
    }


def file_offer(filename: str, record: dict):
    return {
        "type": FILE_OFFER,
        "filename": filename,
        "record": record,
    }


def file_offer_response(filename: str, accepted: bool, message: str = ""):
    return {
        "type": FILE_OFFER_RESPONSE,
        "filename": filename,
        "accepted": accepted,
        "message": message,
    }


def file_chunk(filename: str, data_hex: str, record: dict, done: bool = True):
    return {
        "type": FILE_CHUNK,
        "filename": filename,
        "data": data_hex,
        "record": record,
        "done": done,
    }


def key_migration(new_pub: str, old_sig: str, new_sig: str):
    return {
        "type": KEY_MIGRATION,
        "new_pub": new_pub,
        "old_sig": old_sig,
        "new_sig": new_sig,
    }


def error_message(message: str):
    return {
        "type": ERROR,
        "message": message,
    }


def encode_payload(message: dict) -> bytes:
    return json.dumps(message, separators=(",", ":"), sort_keys=True).encode("utf-8")


def decode_payload(payload: bytes) -> dict:
    return json.loads(payload.decode("utf-8"))
